tasks/ably/ably.go

```diff
@@ -4,23 +4,332 @@ import (
+						return
+					}
+					m := &ablyrpc.Message{}
+					if err := m.UnmarshalJSON(msg.Data); err != nil {
+						log.Error("unmarshalling message", "err", err)
+						boomer.RecordFailure("ably", "subscribe", 0, err.Error())
+					}
+					validateMsg(m, log)
+				}); err != nil {
+					wg.Done()
+					if ctx.Err() != nil {
+						return
+					}
+					log.Error("subscribing", "err", err, "num", i+1)
+					boomer.RecordFailure("ably", "subscribe", 0, err.Error())
+				}
+			}()
+		}
+	}
+	return subClients, nil
+}
+
+func (t *Task) createAblySubscribers(ctx context.Context, channelName string, wg sync.WaitGroup, errorChannel chan error) ([]io.Closer, error) {
+	log := t.conf.Logger
+	subClients := make([]io.Closer, 0, t.conf.NumSubscriptions)
+	for i := 0; i < t.conf.NumSubscriptions; i++ {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+			log.Info("creating subscriber realtime connection", "num", i+1)
+			subClient, err := newAblyClient(t.conf.APIKey, t.conf.Env)
+			if err != nil {
+				subClient.Close()
+				return nil, fmt.Errorf("creating  %dth subscriber; %w", i+1, err)
+			}
+
+			channel := subClient.Channels.Get(channelName)
+
+			log.Info("creating subscriber", "num", i+1, "channel", channelName)
+			sub, err := channel.Subscribe()
+			if err != nil {
+				subClient.Close()
+				return nil, fmt.Errorf("subscribing to %dth channel; %w", i+1, err)
+			}
+
+			wg.Add(1)
+			go t.reportSubscriptionToLocust(ctx, sub, subClient.Connection, errorChannel, &wg, log.New("channel", channelName))
+			subClients = append(subClients, subClient)
+		}
+	}
+	return subClients, nil
+}
+
+func generateChannelName(numChannels, number int) string {
+	return "test-channel-" + strconv.Itoa(number%numChannels)
+}
+
 func newAblyClient(apiKey, env string) (*ably.RealtimeClient, error) {
 	options := ably.NewClientOptions(apiKey)
 	options.Environment = env
 
 	return ably.NewRealtimeClient(options)
 }
 
-func reportSubscriptionToLocust(
+func millisecondTimestamp() int64 {
+	nanos := time.Now().UnixNano()
+	millis := nanos / int64(time.Millisecond)
+	return millis
+}
+
+func (t *Task) randomString(length int) string {
+	b := make([]rune, length)
+	for i := range b {
+		b[i] = t.letters[rand.Intn(len(t.letters))]
+	}
+	return string(b)
+}
+
+func randomDelay(log log15.Logger) {
+	r := rand.Intn(60)
+	log.Info("introducing random delay", "seconds", r)
+	time.Sleep(time.Duration(r) * time.Second)
+	log.Info("continuing after random delay")
+}
+
+func (t *Task) publishOnInterval(
+	ctx context.Context,
+	publishInterval,
+	msgDataLength int,
+	channel *ably.RealtimeChannel,
+	delay int,
+	errorChannel chan<- error,
+	wg *sync.WaitGroup,
+	log log15.Logger,
+) {
+	log = log.New("channel", channel.Name)
+	log.Info("creating publisher", "period", publishInterval)
+
+	log.Info("introducing random delay before starting to publish", "seconds", delay)
+	time.Sleep(time.Duration(delay) * time.Second)
+	log.Info("continuing after random delay")
+
```

Comment:
add a `defer wg.Done()` rather than calling it on every return.

Reply:
Fixed in 7514d46.